Editor refactoring

map-per opened the pull request:
fixes  #9129, fixes #8102, fixes #2298, fixes #6260, fixes #6292, fixes #9126, fixes #6939, fixes #9508, fixes #9311

This PR supersedes #9138 and adds a Journal to explicitly store which changes were submitted by the user. See #9129 for why his information is needed when uploading to OSM.

#### Approach
Edits are logged in the Journal. A log contains info like e.g. `Tag "website" changed from "www.organicmaps.app" to "organicmaps.app"` or `ObjectCreated, Type amenity-bench, Location ...`. When uploading, only the Journal entries are used do update the OSM object. After a successful upload, the Journal is cleared.

Saving and restoring the Journal after an app restart is implemented now.

For the transition phase to the new editing logic the old uploading code is still functional. The old code is used when the object was modified in a previous OM version.

#### Todo
- [x] Logging
    - [x] Add logging for all Tags
    - [x] Add logging for FeatureType
    - [x] Safe and Restore Data after App restart
- [x] New uploading logic
- [x] Transition from old OM versions

#### Bugs fixed
- #8102
- #2298
- #6260
- #6292
   - #9126

#### Testing
Tested with the OSM dev API, but only on Android.

#### Transition
The PR only affects new edits. Old edits from previous OM versions that were not uploaded yet use the old uploading code.

#### New features I keep in mind to not block anything
- Adding "komplex" Category types (#4523)
- Mark POI as disused
- POI category change (#3491)
- Add address POIs (#9199)
- Support editing POIs mapped as areas (#6727)

#### Follow up issues (Issues not addressed in this PR)
- #9043
- #9503
- Remove code of the old editor after some time (e.g. three months)

map-per commented:
@biodranik Can you please take a look to make sure this goes into the right direction.

map-per commented:
This PR is mostly ready. Only the transition logic from old OM versions is missing (I want flag objects edited with old versions with `LegacyObject` entries in the Journal. The uploading logic will then use the old code.). 

Can you please take a look.

map-per commented:
I'm still unsure about how to store the featureType between app restarts. 

I see these options:
- The number (e.g. 16815)
- As a string (e.g. tourism|artwork|)
- With OSM tags (e.g. tourism=artwork)

The code for converting featureType numbers to strings and vice versa is quite complicated and I don't understand where the number comes from (it's not from mapcss-mapping.csv) and whether it is stable or changes with OM updates.

RedAuburn commented:
> I'm still unsure about how to store the featureType between app restarts.

AFAIK the OSM tags are just interpolated from the type string. The number sounds like some internal representation that shouldn't be relied on.

i _think_ it'd be best to store the type string; then it can be passed to 
https://github.com/organicmaps/organicmaps/blob/4c4a8a9784b26d613ad156915da616c3666b0470/indexer/classificator.cpp#L382-L387

map-per commented:
I agree, type string looks best. Will change the code to use it.

Type string is stable and it can be easily converted back to a feature type number. Converting OSM tags back to a feature type number would get quite complex if we want to support more complex types.

map-per commented:
Ok, the PR is finished and tested on Android.

The iOS builds fail though, and I don't know how to fix it. It looks like the compiler doesn't know the new `edit_journal` files. Is there some extra file like `CMakeLists.txt` for iOS?

dvdmrtnz commented:
> Ok, the PR is finished and tested on Android.
> 
> The iOS builds fail though, and I don't know how to fix it. It looks like the compiler doesn't know the new `edit_journal` files. Is there some extra file like `CMakeLists.txt` for iOS?

@map-per Yes, they need to be added to xcode project. Replace `project.pbxproj` in `organicmaps/xcode/indexer/indexer.xcodeproj/project.pbxproj` with the attached file here to make it compile. You will need to extract the file first as I cannot upload `.pbxproj` files directly to github.


[project.pbxproj.zip](https://github.com/user-attachments/files/17356079/project.pbxproj.zip)

map-per commented:
> @map-per Yes, they need to be added to xcode project. Replace project.pbxproj in organicmaps/xcode/indexer/indexer.xcodeproj/project.pbxproj with the attached file here to make it compile. You will need to extract the file first as I cannot upload .pbxproj files directly to github.

Thank you! It worked!

vng commented:
Did quick test on Android:
- Added shop=kiosk or shop=pet, got tags shop=shop, shop=shop ?!
- Added tourism=gallery, got tags tourism=tourism.

Meanwhile, POIs are displayed ok with the correct type and icons. We send bad tags into OSM.

map-per commented:
I can reproduce it. For some reason it only works in the debug version and not in the FDroid version.

map-per commented:
I fixed the problem with the assertion, the F-Droid version works now.

vng commented:
- When I add some POI that is already present in OSM (was added manually or with another app), the POI's icon disappears after the sync try (I suspect so). Not a big deal, but anyway.
- What is Editor's Journal (implemented here) and how can I check it?

map-per commented:
> Well, Android Editor from this PR adds duplicating POIs

Are you sure? To me https://www.openstreetmap.org/changeset/157857343 rather looks like you intentionally created duplicates. Not overwriting existing OSM features is intentional (#2298), it's better to add a new node in such cases instead of adding the tags to some random existing feature.

> What is Editor's Journal (implemented here) and how can I check it?

The journal is implemented in `edit_journal.cpp/.hpp`. The debug version prints what changes were added to the journal. It also prints the full journal before saving in xml, after loading form xml and before editing OSM. In Logcat I use "package:mine (indexer/edit | editor/ | feature_data)" to filter for journal related debug info.

> When I add some POI that is already present in OSM (was added manually or with another app), the POI's icon disappears after the sync try (I suspect so). Not a big deal, but anyway. Did a test on iOS.

That should not happen. Is the icon entirely gone or just hidden/displaced by a different icon?

map-per commented:
Looks like iOS updates the core every time the input text changes and not, like Android, when saving. That's a problem because every change generates a new journal entry.

vng commented:
Yes, I intentionally created duplicates to check how this PR handles it.
Before this PR - duplicates were not uploaded to OSM, this PR - they are uploaded.

"Create new node" when tags are different is ok, but why make duplicates?

map-per commented:
I analyzed this in https://github.com/organicmaps/organicmaps/issues/2298#issuecomment-2308788517. For 100 edits, the current approach worked correctly for 75. In 16 other objects were wrongly overwritten resulting in things like `natural=tree` + `amenity=bench`. In 9 edits the merging was in general a good thing to do, but OM doesn't properly remove old tags, leading to wrong tagging like `amenity=bakery` + `office=tax_advisor`.

- In this PR the edits form the 75 category as well as the edits from the 16 category are handled correctly. The edits form the 9 category, however result in duplicates. So, over all, more edits are correct in this PR than previously.

- Moreover duplicate nodes are easier to see and fix for OSMers than wrongly merged nodes. So it's better to error on this side.

- And, maybe not so important: The behavior of OM is more transparent. The user clicks "Add a place" and a new node is created. So with this PR OSM reviewers can ask OM users to be careful not to add duplicates, but use the "This place doesn't exist" option instead. With the current approach OM users can't really do anything to prevent their edits form being merged wrongly. For the mentioned tree example the tree is not even displayed on the map.

RedAuburn commented:
(i've got this PR on my phone now, and will try to make as many edits as i can to test 👍️)

map-per commented:
The new version does the journal Logging later, just before the changes are saved in the storage. This might fix the iOS problems.

For now only creating objects and editing of Metadata values (e.g. opening_hours, phone, operator, website, email, postcode, building_levels, social media) is supported. Deleting tags or changing e.g. cuisine or names is not implemented.

@vng Can you please test it on iOS?
In case it doesn't work it would be good if you could describe the problem in more detail. Does editing work, does loading the data after an App restart work, does uploading to OSM work, does the app crash?

vng commented:
I still want to have some duplicating filtering at least like the current master version, e.g:
| Existing POI | Adding POI  |Result in OSM|
| ------------ | ------------| --------------- |
|shop=kiosk   | shop=kiosk | Nothing added |
|name=XXX    |
|-------------|------------ |-----------------|
|shop=kiosk   | shop=kiosk | Have 2 POIs with different names | 
|name=XXX    | name=YYY | Probably add Note |

map-per commented:
In general I agree that a merge algorithm is nice to have at some point. But that is outside of the scope of this PR. You can't expect me to fix all OM editor problems at once, reducing the number of wrong changesets should be enough.

Please don't overestimate the importance of your example. From my changeset analysis I can tell that adding a `shop=kiosk` on top of a `shop=kiosk` is not something OM users are actually doing. People only add different POI types on top of others to "change the category". (An with the visual feedback that OM gives them this appears to actually works.)

A merge algorithm would be more interesting for cases where duplicates are caused by outdated map data. But in these cases it's unlikely that users, by change, hit within one meter of the existing POI. So these edits are not affected by the behavior change in this PR. (To tackle this problem we would need to search in a bigger radius.)

The overwriting/new node change will affect about 120 changesets a month, so the harm to OSM would be very limited in case my opinion and my analysis turned out to be wrong. I would suggest to change the behavior and add an `info:feature_close_by=yes` flagging. Then we can check and measure the success afterwards.

(I don't want this rather minor change to block the whole PR. So if you insist on not changing something about #2298 I can also restore the old behavior. IMO that would be stupid though.)

vng commented:
I'm not going into Merge function, I understand how difficult it is to implement.
I have only 3 outcomes: add / skip / add with an additional note (maybe).

map-per commented:
Come one, can we please talk about real problems instead of making mountains out of case that never actually happens. I'm getting really frustrated by now.

vng commented:
I'm sorry to hear you're feeling frustrated. I completely understand how challenging this task is and how much effort you made here. Ok, take my comments as "nice-to-have", but not like "must-have" actions.

map-per commented:
I thought it would be good to mention what editor issues are a priority for me: I collect the most relevant editor bugs in #6364 and UX issues in #9385. So these are the list I'm currently going through.

map-per commented:
Multiple uploads of POIs (#2071) might cause problems in combination with #2298. This bug is not harmful at the moment because the second upload of a POI is merged into the node that was created in the first upload. With the new behavior, however, #2071 results in duplicate nodes. 

We should keep an eye on this when doing the Alpha testing. Depending on how often #2071 occurs it might be better to remove the #2298 fix form this PR.

vng commented:
Added shop: https://www.openstreetmap.org/node/12291931301 with addr:street + addr:housenumber.
See address in OM, but it was not uploaded into OSM.

vng commented:
Edit POI, set empty name (like should be) and OSM already has an empty name (was edited by someone else).
Got:
```
editor/osm_editor.cpp:713 operator()(): CHECK(false) Object already IN_SYNC should not be here
```

vng commented:
Duplicating notes are still exist:
https://www.openstreetmap.org/note/4497044
https://www.openstreetmap.org/note/4497035

map-per commented:
Thank you for the review!
The greengrocer and the bar worked as expected. Since the last commit only metadata fields are supported (see: https://github.com/organicmaps/organicmaps/pull/9242#issuecomment-2423908939).

The crash is a bug, thanks for the detailed description, I need to take a closer look at why this happens. (It looks like there is a loophole that allows saving unchanged objects in the internal storage.)

I did not change any OSM Note related code.


Did / can you also test iOS? iOS is what I'm most interested in, because I'm still quite puzzled what "Probably, iOS Editor doesn't work at all .." means. It would be good if you could describe the problem in more detail. Does editing work, does loading the data after an App restart work, does uploading to OSM work, does the app crash? And please only test with metadata fields (https://github.com/organicmaps/organicmaps/pull/9242#issuecomment-2423908939).

map-per commented:
The new version is now fully functional again. I moved code that I accidentally added to the Android only Editor.cpp file, so this might fix the bugs on iOS.

 To avoid duplicated nodes after network failures (https://github.com/organicmaps/organicmaps/pull/9242#issuecomment-2438168824), I added a check that merges features at the exact same location (no 1m tolerance any more) (tested with https://master.apis.dev.openstreetmap.org/node/4352709181/history)

map-per commented:
I tested with the real OSM database and did some code clean up. Now this is ready for review.

map-per commented:
Yes, the version is fully functional, so addresses and names work

vng commented:
Don't know if it is connected with refactoring or not, but:
"Request Network error while connecting to https://api.openstreetmap.org/api/0.6/node/create"

https://www.openstreetmap.org/changeset/159418301

Did it with fast and stable in-house WiFi.

map-per commented:
The network error issue is tracked in #8552, so it's not caused by the refactoring.

vng commented:
I suspect that something happened with the changes upload trigger.
Recent changeset https://www.openstreetmap.org/changeset/159553859

POIs were added yesterday and today several hours ago. The app has been in the background for a long time since yesterday. Changeset has been created only after the manual Settings -> "Profile OpenStreetMap".

map-per commented:
I did not change the upload trigger code, so I don't think this is related to this PR.

I don't really know the upload trigger code, but I think OM does not try to upload in the background. Uploads only happen during the transition from foreground to background.

map-per commented:
I tested it in the Playstore version: Edits are not uploaded in the background, but only at the transition from foreground to background

map-per commented:
So we both tested a few days and found no (PR related) bugs:

Can we ship the PR to Android Firebase? Or should we first test iOS?

vng commented:
> I tested it in the Playstore version: Edits are not uploaded in the background, but only at the transition from foreground to background

Exactly. Edits are uploaded immediately after I turn off/switch off the app on master. But It didn't happen with this PR. Will check once more, maybe OSM Api had lags.

Please, also check yourself.

map-per commented:
That doesn't really make sense a I did not change the upload trigger code. I suspect that there was no internet connection when you first closed the app.

vng commented:
Did some testing today - uploading is ok.

@map-per Do you want to make any additional commits reorganization? Actually, 67 commits are ok for me. Anyway, I'm going to make a Merge commit here but keep the full history.

Will do a code review later.

map-per commented:
Initially I thought about squashing everything into one commit, but keeping the development history is also a good idea. So let's just stick to the 67 commits.

vng commented on editor/osm_editor.cpp:
```diff
@@ -824,9 +980,14 @@ bool Editor::FillFeatureInfo(FeatureStatus status, XMLFeature const & xml, Featu
+      editor::ApplyPatch(xml, fti.m_object);
   }
 
+  fti.m_object.SetJournal(journal);
```
std::move(journal)

vng commented on indexer/editable_map_object.cpp:
```diff
@@ -578,6 +579,209 @@ bool EditableMapObject::ValidateName(string const & name)
+  return journal;
+}
+
+void EditableMapObject::SetJournal(EditJournal editJournal)
```
EditJournal && editJournal

vng commented on indexer/editable_map_object.cpp:
```diff
@@ -578,6 +579,209 @@ bool EditableMapObject::ValidateName(string const & name)
+
+void EditableMapObject::SetJournal(EditJournal editJournal)
+{
+  journal = editJournal;
```
std::move(editJournal)

vng commented on editor/xml_feature.cpp:
```diff
@@ -409,6 +409,135 @@ string XMLFeature::GetUploadError() const { return GetRootNode().attribute(kUplo
+          entry.data = legacyObjData;
+          break;
+        }
+      }
```
What will happen if xml file is broken, say any XXX.value() does not exist? Should we have any generic error processing here? Like catch an exception and continue/skip the whole entry?

map-per replied:
pugixml shouldn't throw exceptions. It's very fault tolerant, so e.g. calling `.attribute("key").value()` on an empty xml node doesn't crash, but just returns "".

The current `GetEditJournal()` assumes that the xml is correct. Should I add checks and use MYTHROW in case read in values are invalid like it is done in other parts of the `xml_feature` class?

map-per replied:
I added checks to make sure only correct values are read from the xml file. (exceptions are thrown when the xml contains an invalid value)

vng commented on editor/xml_feature.cpp:
```diff
@@ -409,6 +409,135 @@ string XMLFeature::GetUploadError() const { return GetRootNode().attribute(kUplo
+    // Mark as Legacy Object
+    osm::LegacyObjData legacyObjData = {"1.0"};
+    journal.AddJournalEntry({osm::JournalEntryType::LegacyObject, time(nullptr), legacyObjData});
+    journal.AddJournalHistoryEntry({osm::JournalEntryType::LegacyObject, time(nullptr), std::move(legacyObjData)});
```
Call time(nullptr) once here?

vng commented on editor/xml_feature.cpp:
```diff
@@ -449,6 +578,42 @@ void XMLFeature::RemoveTag(string_view key)
 
+void XMLFeature::UpdateOSMTag(std::string_view key, std::string_view value)
+{
+  if (value == "") {
```
value.empty()

vng commented on editor/xml_feature.cpp:
```diff
@@ -611,6 +776,36 @@ XMLFeature ToXML(osm::EditableMapObject const & object, bool serializeType)
+  }
+  else
+  {
+    string const strType = classif().GetReadableObjectName(uint32_t(type));
```
```GetReadableObjectName(type)```

vng commented on editor/xml_feature.cpp:
```diff
@@ -611,6 +776,36 @@ XMLFeature ToXML(osm::EditableMapObject const & object, bool serializeType)
+
+    CHECK(++iter, ("Processing Type failed: ", strType));
+    // Main type is always stored as "k=amenity v=restaurant".
+    toFeature.SetTagValue(k, *iter);
```
I suspect that we have more 3-arity types besides amenity-recycling-xxx
Add ASSERT/CHECK(!(++iter), ("Add an exception:", strType))

map-per replied:
:+1: Added an assertion

vng commented on indexer/edit_journal.hpp:
```diff
@@ -0,0 +1,84 @@
+  class EditJournal
+  {
+    std::list<JournalEntry> journal{};
+    std::list<JournalEntry> journalHistory{};
```
Please, use "m_" notation at least here.

vng commented on indexer/edit_journal.cpp:
```diff
@@ -0,0 +1,124 @@
+    }
+    else if (journal.front().journalEntryType == JournalEntryType::ObjectCreated) {
+      return EditingLifecycle::CREATED;
+    }
```
Braces are not needed here

vng commented on indexer/edit_journal.cpp:
```diff
@@ -0,0 +1,124 @@
+    TagModData tagModData = {key, old_value, new_value};
+    JournalEntry entry = {JournalEntryType::TagModification, time(nullptr), std::move(tagModData)};
+    AddJournalEntry(entry);
+    LOG(LDEBUG, ("Key ", key, "changed from \"", (std::string) old_value, "\" to \"", (std::string) new_value, "\""));
```
Here is how I'd prefer to see this code:
```
LOG(LDEBUG, ("Key ", key, "changed from \"", old_value, "\" to \"", new_value, "\""));
AddJournalEntry({JournalEntryType::TagModification, time(nullptr), {key, std::move(old_value), std::move(new_value)}});
```

vng commented on indexer/edit_journal.cpp:
```diff
@@ -0,0 +1,124 @@
+    LOG(LDEBUG, ("Object of type ", classif().GetReadableObjectName(type), " created"));
+  }
+
+  void EditJournal::AddJournalEntry(JournalEntry const & entry)
```
JournalEntry entry

vng commented on indexer/edit_journal.cpp:
```diff
@@ -0,0 +1,124 @@
+
+  void EditJournal::AddJournalEntry(JournalEntry const & entry)
+  {
+    journal.push_back(entry);
```
std::move(entry)

vng commented on indexer/edit_journal.cpp:
```diff
@@ -0,0 +1,124 @@
+  {
+    for (JournalEntry const & entry : journal) {
+      journalHistory.push_back(entry);
+    }
```
```
for (JournalEntry & entry : journal)
  journalHistory.push_back(std::move(entry));
```

vng commented on indexer/edit_journal.cpp:
```diff
@@ -0,0 +1,124 @@
+    return journalHistory;
+  }
+
+  void EditJournal::AddJournalHistoryEntry(JournalEntry const & entry)
```
JournalEntry entry

vng commented on indexer/edit_journal.cpp:
```diff
@@ -0,0 +1,124 @@
+
+  void EditJournal::AddJournalHistoryEntry(JournalEntry const & entry)
+  {
+    journalHistory.push_back(entry);
```
std::move(entry)

vng commented on indexer/edit_journal.cpp:
```diff
@@ -0,0 +1,124 @@
+        return "ObjectCreated";
+      case osm::JournalEntryType::LegacyObject:
+        return "LegacyObject";
+    }
```
UNREACHABLE();

map-per replied:
Is it UNREACHABLE() necessary? All three JournalEntryType enum values are handled here, so the compiler can figure out by itself that the end is unreachable.

vng commented on indexer/edit_journal.cpp:
```diff
@@ -0,0 +1,124 @@
+    ObjCreateData objCreateData = {type, geomType, std::move(mercator)};
+    JournalEntry entry = {JournalEntryType::ObjectCreated, time(nullptr), std::move(objCreateData)};
+    AddJournalEntry(entry);
+    LOG(LDEBUG, ("Object of type ", classif().GetReadableObjectName(type), " created"));
```
```
ASSERT(journal.empty(), ("Only empty journals can be marked as created"));
LOG(LDEBUG, ("Object of type ", classif().GetReadableObjectName(type), " created"));
AddJournalEntry({JournalEntryType::ObjectCreated, time(nullptr), {type, geomType, mercator}});
```

vng commented on indexer/editable_map_object.cpp:
```diff
@@ -578,6 +579,209 @@ bool EditableMapObject::ValidateName(string const & name)
+  }
+  for (JournalEntry const & entry : editJournal.GetJournal()) {
+    ApplyJournalEntry(entry);
+  }
```
No need braces here

vng commented on indexer/editable_map_object.cpp:
```diff
@@ -578,6 +579,209 @@ bool EditableMapObject::ValidateName(string const & name)
+
+      //Names
+      uint8_t langCode = StringUtf8Multilang::GetCodeByOSMTag(tagModData.key);
+      if (langCode != 255) { //not UnsupportedLanguageCode
```
Ok, so check ```langCode != StringUtf8Multilang:: UnsupportedLanguageCode```

vng reviewed (commented):
Will finish review later.

Please:
- setup your editor to use open { braces from the new line (not KR-style)
- don't use braces for one-line expressions

map-per commented on android/app/src/main/cpp/app/organicmaps/editor/Editor.cpp:
```diff
@@ -149,7 +149,8 @@ Java_app_organicmaps_editor_Editor_nativeHasWifi(JNIEnv *, jclass)
 Java_app_organicmaps_editor_Editor_nativeSetHasWifi(JNIEnv *, jclass, jboolean hasWifi)
 {
-  g_editableMapObject.SetInternet(hasWifi ? feature::Internet::Wlan : feature::Internet::Unknown);
+  if (hasWifi != (g_editableMapObject.GetInternet() == feature::Internet::Wlan))
```
That's not a good idea, because iOS uses all `Internet` values and not just Wifi `true`/`false`.
(This is a fix for an Android only bug)

vng replied:
ok

map-per commented on coding/string_utf8_multilang.cpp:
```diff
@@ -149,6 +149,32 @@ std::vector<std::string_view> const * StringUtf8Multilang::GetTransliteratorsIds
+{
+  std::string_view lang = StringUtf8Multilang::GetLangByCode(langCode);
+  if (lang == "")
+    return "";
```
Added ASSERT_FAIL for that case

map-per commented on editor/osm_editor.cpp:
```diff
@@ -613,110 +624,242 @@ void Editor::UploadChanges(string const & oauthToken, ChangesetTags tags,
+                          ASSERT_FAIL(("Legacy Objects can not be edited with the new editor"));
+                          break;
+                      }
+                    }
```
Moved it to a new function

map-per commented:
Thanks for the first part of the review!
I included the suggestions and fixed the brace code style.

biodranik commented:
> Initially I thought about squashing everything into one commit, but keeping the development history is also a good idea. So let's just stick to the 67 commits.

We can keep a copy of the original branch with all intermediate commits separately, just in case you need it.

I prefer a few logically organized commits for easier reverting/improving.

vng commented on editor/osm_editor.cpp:
```diff
@@ -613,110 +624,213 @@ void Editor::UploadChanges(string const & oauthToken, ChangesetTags tags,
-              if (osmFeature == osmFeatureCopy)
+            LOG(LDEBUG, ("New Editor used\n"));
+
+            switch (fti.m_status) {
```
Please, all braces { here and below from the new line.

vng commented on indexer/edit_journal.cpp:
```diff
@@ -0,0 +1,125 @@
+  void EditJournal::AddTagChange(std::string key, std::string old_value, std::string new_value)
+  {
+    LOG(LDEBUG, ("Key ", key, "changed from \"", old_value, "\" to \"", new_value, "\""));
+    AddJournalEntry({JournalEntryType::TagModification, time(nullptr), TagModData{key, std::move(old_value), std::move(new_value)}});
```
std::move(key)

vng commented on indexer/editable_map_object.cpp:
```diff
@@ -578,6 +579,217 @@ bool EditableMapObject::ValidateName(string const & name)
+    if (new_name != old_name)
+    {
+      std::string osmLangName = StringUtf8Multilang::GetOSMTagByCode(langCode);
+      m_journal.AddTagChange(osmLangName, std::string(old_name), std::string(new_name));
```
std::move(osmLangName)

vng commented on indexer/editable_map_object.cpp:
```diff
@@ -578,6 +579,217 @@ bool EditableMapObject::ValidateName(string const & name)
+  {
+    auto const type = static_cast<feature::Metadata::EType>(i);
+    std::string const & value = std::string(GetMetadata(type));
+    std::string const & old_value = std::string(unedited_emo.GetMetadata(type));
```
std::string_view value = ...
std::string_view old_value = ...

vng commented on indexer/editable_map_object.cpp:
```diff
@@ -578,6 +579,217 @@ bool EditableMapObject::ValidateName(string const & name)
+    std::string const & old_value = std::string(unedited_emo.GetMetadata(type));
+
+    if (value != old_value)
+      m_journal.AddTagChange(ToString(type), old_value, value);
```
m_journal.AddTagChange(ToString(type), std::string(old_value), std::string(value));

vng commented on indexer/editable_map_object.cpp:
```diff
@@ -578,6 +579,217 @@ bool EditableMapObject::ValidateName(string const & name)
+  std::vector<std::string> new_cuisines = GetCuisines();
+  std::vector<std::string> old_cuisines = unedited_emo.GetCuisines();
+
+  auto const findAndErase = [] (std::vector<std::string>* cuisinesPtr, std::string s)
```
```(std::vector<std::string> & cuisinesPtr, std::string_view s)```

vng commented on indexer/editable_map_object.cpp:
```diff
@@ -578,6 +579,217 @@ bool EditableMapObject::ValidateName(string const & name)
+          goto value_found;
+      }
+      cuisinesModified = true;
+      value_found:;
```
```
if (!base::IsExist(old_cuisines, new_cuisine))
{
  cuisinesModified = true;
  break;
}
```

map-per replied:
Thanks! That makes the code way cleaner than with the `goto`

map-per commented:
Thanks, I included the second part of the review.

> I prefer a few logically organized commits for easier reverting/improving.

The PR can not really be divided into several sub-changes because things only work with all changes. So this would mean merging everything into one commit.

map-per commented:
Should I squash all commits or should I not touch the already reviewed commits?

map-per commented:
A list of interesting cases for testing on iOS:

- Editing POI works, changes show up on the map
- Adding POI works, changes show up on the map
- Changes are still there after completely closing the app
- Changes are uploaded to OSM

map-per commented:
Let's communicate a little more. Are you planning to do the iOS testing yourself once you find the time for it or should I ask other iOS developers for help?

vng commented:
I built this branch rebase on latest master.

vng commented:
I built this branch, rebase on the latest master for iOS:
- I can't log in my OSM account (the current app store app version does)
- Here is the screenshot of the invalid category search:

![image](https://github.com/user-attachments/assets/b922d7f2-649e-4ed6-a46d-91d05a705914)


CC @kirylkaveryn

kirylkaveryn commented:
> CC @kirylkaveryn

Hmm... I'll take a look

vng commented:
@kirylkaveryn  Check the UI only, seems like login is invalid for the Debug DevServer.
@map-per Next launch crash on
```
bool loadFromJournal = journal.GetJournalHistory().front().journalEntryType != JournalEntryType::LegacyObject;
```
GetJournalHistory() is empty.

map-per commented:
Thank you for testing!

I fixed the bug and added checks for empty journals in two places.

> - I can't log in my OSM account (the current app store app version does)

The OM debug versions use the OSM dev server (https://master.apis.dev.openstreetmap.org). So the login should work when you use / create a new account for this database. I e.g. use https://master.apis.dev.openstreetmap.org/user/map%20per%20dev for testing.

> - Here is the screenshot of the invalid category search:

@kirylkaveryn Did you see #9812?

vng commented on editor/osm_editor.cpp:
```diff
@@ -935,7 +936,9 @@ bool Editor::FillFeatureInfo(FeatureStatus status, XMLFeature const & xml, Featu
   // Do not load Legacy Objects form Journal
-  bool loadFromJournal = journal.GetJournalHistory().front().journalEntryType != JournalEntryType::LegacyObject;
+  auto const & journalHistory = journal.GetJournalHistory();
+  bool loadFromJournal = journalHistory.empty() || journalHistory.front().journalEntryType != JournalEntryType::LegacyObject;
```
Tale out into a separate function to avoid copy-paste-check ?

map-per replied:
It's not really worth it because I want to remove the decision logic for old/new editor in a few months anyway. This code is only needed for the transition period.

kirylkaveryn commented:
> @kirylkaveryn Did you see #9812?

fixed here: https://github.com/organicmaps/organicmaps/pull/9915

map-per commented:
Let's continue. The undefined behavior is fixed, so iOS should work now.

map-per commented:
@organicmaps/mergers 
How can we overcome the review bottleneck and finalize this PR? The PR is ready, it just needs final reviews and testing. Let's finally fix the ~9 issues addressed in this PR.

_(Rebased the branch and squashed the commits. The old history is in https://github.com/map-per/organicmaps/tree/2025-01-31_editor_refactoring)_

vng reviewed (approved):
I did permanent testing of this PR for 2 months. Merge, and everyone interested can check it in Beta.

RedAuburn commented:
Congrats @map-per 🎉 thank you for all your hard work on this PR ❤️
